### PR TITLE
bugfix: render and hit test portals in the correct order when they are added in different frames

### LINKED ISSUE #15: Having trouble using a Portal inside of a Portal to recreate a 'Dialog' triggered from a 'Modal'
I was having a lot of fun creating custom Datepickers and Dropdown buttons for myself, which work great.
Then I made a kind of modal/form which is also a PortalEntry, which works great.
But my problem is when I try to use one of my Datepickers/Dropdowns inside of my Pop up form, I get the following:

LeaderLayer anchor must come before FollowerLayer in paint order, but the reverse was true.
'package:flutter/src/rendering/layer.dart':

I assume this is because I am putting the PortalEntry inside of a portal, not a PortalEntry child. But I am struggling to find a workaround.

I tried putting another portalEntry in the form child and then passing the next portal I want when the corresponding button is clicked in the form. But the portal passed shows up underneath the form.

I tried even just showing a regular Dialog from the form. But it shows up underneath.

Would really appreciate any tips. Also just want to say, awesome plugin, allows for some really nice custom widgets

### PR DESCRIPTION
**Pull request by shrouxm**
hello! i think i have a fix for #15. the first commit adds a test which demonstrates the crash from #15, and the second commit is the fix. the updated code passes all existing tests

explanation: in the original code, multiple portals which are layed out during the same frame will have their overlays added to the overlay link in reverse order, so they're need to be iterated in reverse order when rendering. but a new portal added in a subsequent frame will therefore get rendered *first* instead of last. this PR fixes this behaviour with the following logic in _OverlayLink: we maintain two sets of overlays, '_overlays' and '_newOverlays'. we only ever add to '_newOverlays', and whenever accessing '_overlays' we first add the '_newOverlays' in reverse order before emptying '_newOverlays'. this way overlays added during the same frame still get added in reverse order, but an overlay added in a later frame will come after. the assumption now is that '_overlays' is not in reverse order, so we reverse the order of iteration when accessing it. i also made the sets LinkedHashSets explicitly to make clear that the iteration in insertion order is required.

i'm pretty new to flutter so i'm not sure if there's any hidden drawbacks to this method, or i've implemented things in an awkward way. in particular, i wasn't sure how to make sure '_newOverlays' gets flushed to '_overlays' exactly once per frame and instead do it every time '_overlays' is accessed which seems like it might have a subtle edge case.

fixes #15

### CODE REVIEW

File: test/widget_test.dart [hunk 1]
```diff
@@ -1571,6 +1571,57 @@ Future<void> main() async {
...
+                ),
+                child: Container(
+                  height: 30,
+                  width: 30,
+                  color: Colors.red,
+                ),
+              ),
+            ),
+            child: Center(
+              child: Container(
```
Comment (rrousselGit) [inline test/widget_test.dart:1607]: Why is this container on the top left in the golden test if it is centered?

File: lib/src/portal.dart [hunk 2]
```diff
@@ -142,16 +158,15 @@ class _RenderPortalTheater extends RenderProxyBox {
...
-
+    for (final overlay in overlayLink.overlays) {
       context.paintChild(overlay, offset);
     }
   }
 
   @override
   bool hitTestChildren(BoxHitTestResult result, {required Offset position}) {
-    for (final overlay in overlayLink.overlays) {
+    for (var i = overlayLink.overlays.length - 1; i >= 0; i--) {
```
Comment (rrousselGit) [inline lib/src/portal.dart:168]: What is the purpose of this change?

File: lib/src/portal.dart [hunk 3]
```diff
@@ -69,7 +70,22 @@ class _OverlayLink {
...
+  final LinkedHashSet<RenderBox> _newOverlays = LinkedHashSet.identity();
+
+  void addOverlay(RenderBox overlay) => _newOverlays.add(overlay);
+  void removeOverlay(RenderBox? overlay) {
+    _overlays.remove(overlay);
+    _newOverlays.remove(overlay);
+  }
+
+  Iterable<RenderBox> get overlays {
+    while (_newOverlays.isNotEmpty) {
```
Comment (rrousselGit) [inline lib/src/portal.dart:83]: Don't do that in the getter

This logic should be in a separate place

**Review by rrousselGit — CHANGES_REQUESTED** [review]
The `overlays` logic needs a bit of refactoring

**Comment by rrousselGit** [comment]
Also, could you document this approach in the code?

> this way overlays added during the same frame still get added in reverse order, but an overlay added in a later frame will come after. the assumption now is that '_overlays' is not in reverse order, so we reverse the order of iteration when accessing it. i also made the sets LinkedHashSets explicitly to make clear that the iteration in insertion order is required.



As this approach isn't very natural. We'd need a comment to explain what is happening

**Comment by hamirshekhawat** [comment]
@shrouxm  Can you please update this PR?

**Comment by shrouxm** [comment]
@hamirshekhawat i stopped using this library because i ran into other issues which i was unable to resolve, so i stopped working on the PR as well, sorry! i'm happy for anyone else running into the bug to take over the PR though
